**Daniel Goings<br>
dcgoings@buffalo.edu<br>
MTH337 Project 1 - Prime Numbers**

In [119]:
import math
import operator
import collections

# PART 1

The first function, **listPrimes**, accepts an integer *n* as the argument and returns all prime numbers that are smaller than *n* in a list. This function was implemented using the Sieve of Eratosthenes. 

To accomplish this, we first initiate a list of length *n*, all equal to *True*, and begin our for loop. The for loop looks for the first *True* value in the list, which indicates that number is prime, adds that number to our listOfPrimes, and finally sets all multiples of that number to *False* in the original list of all numbers.

In [120]:
def listPrimes(n):
    
    #initialize a list of length n
    allNums = n * [True]
    
    #initialize empty list for our found prime numbers
    listOfPrimes = []

    print("\nlistPrimes returns all prime numbers less than the given input, n.")
    print("n = %s" % n)
    print("Therefore, we will return all prime numbers less than %s\n" % n)
    
    #for loop, from our starting prime number 2, up to n
    for x in range (2, n):
        
        #if the number in list 'allNums' in position 'x' is set to 'True', that number is prime and we add to the list
        if allNums[x] == True:
            listOfPrimes.append(x)
            
            #we now set ALL multiples of 'x' to 'False' in 'allNums'
            for y in range(x*2, n, x):
                allNums[y] = False
                
    return (listOfPrimes)

In [121]:
#testcase
listPrimes(21)


listPrimes returns all prime numbers less than the given input, n.
n = 21
Therefore, we will return all prime numbers less than 21



[2, 3, 5, 7, 11, 13, 17, 19]

In [122]:
#testcase
listPrimes(62)


listPrimes returns all prime numbers less than the given input, n.
n = 62
Therefore, we will return all prime numbers less than 62



[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61]

The second function, **nPrimes**, accepts an integer *n* as the argument and returns the first *n* prime numbers. This function was also implemented using the Sieve of Eratosthenes. 

To accomplish this, we use the same overall idea as in the first function with a few changes. Since we don't know what number to stop at when creating our initial list, we create a safety net and overestimate our goal. We can do this by creating a list equal to *n x 100*, and continue forth setting them all to *True*. We then break out of the loop once the length of *listOfPrimes* equals our *n* value.

NOTE: I used *print(listOfPrimes)* because using *return(listOfPrimes)* caused the list to print out vertical and take 100x more page space


In [123]:
def nPrimes(n):

    #initialize a list that overestimates how many numbers we need to reach the n'th prime number
    length = n * 100
    allNums = length * [True]
    
    #initialize empty list for our found prime numbers
    listOfPrimes = []

    print("\nnPrimes returns the first n prime numbers.")
    print("n = %s" % n)
    print("Therefore, we will print the first %s prime numbers\n" % n)

    #same Sieve idea as used above, except now we go with range(2, length) instead of range(2, n)
    for x in range (2, length):
        if allNums[x] == True:
            listOfPrimes.append(x)
            
            #if the length of our list of primes equals n, we have found our answer!
            if len(listOfPrimes) == n:
                print ("%s is the %s prime number!\n" % (x, n))
                print (listOfPrimes)
                
            for y in range(x*2, length, x):
                allNums[y] = False

In [124]:
#testcase
nPrimes(100)


nPrimes returns the first n prime numbers.
n = 100
Therefore, we will print the first 100 prime numbers

541 is the 100 prime number!

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541]


# PART 2

The first function in Part 2, **primeDivisors**, accepts an integer *n* as the argument and returns the prime divisors as the output in a list. 

To accomplish this, we can use some shortcuts. We know we can limit our range to the square root of our number *n*, since any divisor greater than the square root would already be found between 2 and the square root of *n*. We then iterate through all possible factors, checking whether *n mod factor* has a remainder of zero, re-assigning *n* as *n divided by the factor*, until we retrieve our prime divisors.


In [150]:
def primeDivisors(n):
    print("Calculating the prime divisors for %s" % n)
    
    #as stated above, this significantly reduces our time since any divisor greater than sqrt(n) will already be found
    # '+ 1' is to account for smaller numbers such as 4 which would give us a problem otherwise
    sq = math.sqrt(n) + 1
    
    #our starting factor
    factor = 2
    
    #intiialize empty list of divisors
    listOfDivisors = []

    while factor < sq:
        
        #while n % factor has no remainder, re-assign n to new value,
        #keep looping until the prime divisor is obtained (has a remainder), and append
                
        while n % factor == 0:
            n = n / factor
            listOfDivisors.append(factor)
        
        #current factor complete, check next factor
        factor+=1
        
    #accounts for cases where the you have zero prime divisors (1 and itself), and
    #when two prime divisors are found immediately (as with 21 (7*3))
    if n > 1:
        listOfDivisors.append(int(n))
    return (listOfDivisors)


In [151]:
#testcases
primeDivisors(4)

Calculating the prime divisors for 4


[2, 2]

In [152]:
primeDivisors(18)

Calculating the prime divisors for 18


[2, 3, 3]

In [153]:
primeDivisors(21)

Calculating the prime divisors for 21


[3, 7]

In [154]:
primeDivisors(100)

Calculating the prime divisors for 100


[2, 2, 5, 5]

**BONUS:** The bonus function accepts an integer and returns two outputs, the first being a list of all unique prime divisors, and the second being a list of how many times each of those unique prime divisors occurs.

Although there were multiple ways to do this, Python has a Counter function which does exactly that and returns as a dictionary (keys and values). We can then simply print which we want (keys or values) with type cast to list.

**NOTE:** It appears this import is not compatible with Jupyter Notebook, however it does work in Atom (standalone editor).

In [155]:
def primeDivisorsUnique(n):
    #we can save time by running our above function to retrieve our list of prime divisors
    temp = primeDivisors(n)
    print(temp)
    
    #run the imported collections.Counter function on our list
    result = collections.Counter(temp)

    #separate keys and vals
    keys = list(result.keys())
    vals = list(result.values())
  
    print("Divisors: = %s" % keys)
    print("Quantity: = %s" % vals)

In [156]:
primeDivisorsUnique(18)

Calculating the prime divisors for 18
[2, 3, 3]
Divisors: = [2, 3]
Quantity: = [1, 2]


In [157]:
primeDivisorsUnique(100)

Calculating the prime divisors for 100
[2, 2, 5, 5]
Divisors: = [2, 5]
Quantity: = [2, 2]


# PART 3

In Part 3, using the information provided, we created a function that checks for congruency. This is done by utilizing the pow(a,b,c) function and comparing the answer to a % b.

In [158]:
def isCongruent(n):
    for x in range(1, n):
        if pow(x, n, n) != x % n:
            return False
    return True

The first function, *listPrimeLike*, is nearly the same as *listPrimes* except it returns all prime-like numbers less than *n* using the congruence relation check.

For this function, we run a for loop from 2 to our *n* value, and if that number returns *True* for the isCongruent() function, it implies prime-like. 

In [159]:
def listPrimeLike(n):
    
    #initialize empty list for primes
    primeLikes = []
    
    #iterate up to our desired n value
    for x in range(2, n):
        
        #check for congruency, if so, append to list
        if isCongruent(x):
            primeLikes.append(x)

    print(primeLikes)

In [160]:
listPrimeLike(50)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]


Similarly, the second function does the same as our nPrimes function, which returns the first *n* prime-like numbers. The overall function is identical to listPrimeLike, with the exception of our adjusted max range value and if statement to stop when the length of primeLikes has reached our desired number n of prime-like numbers.

In [161]:
def nPrimeLike(n):
    
    #initialize empty list for primes
    primeLikes = []

    #iterate up to our desired n value, since we don't know, we overestimate by n*100
    for x in range(2, (n*100)):
        
        #check for congruency, if so, append to list
        if isCongruent(x):
            primeLikes.append(x)
    
        #check if the length of primeLikes equals our desired number n
        if len(primeLikes) == n:
            print(primeLikes)
            return

    print(primeLikes)

In [162]:
nPrimeLike(50)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229]


In [163]:
nPrimeLike(200)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 561, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1105, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 121

In [164]:
nPrimes(200)


nPrimes returns the first n prime numbers.
n = 200
Therefore, we will print the first 200 prime numbers

1223 is the 200 prime number!

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 105

Comparing the above two functions, we can see that two numbers show up in *nPrimeLike* that do not show up in *nPrimes*: 561 and 1105. We can conclude that these numbers actually satisfy and pass the congruency relation check, even though they are not prime numbers; hence, **"prime-like"**.